server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/LongRareTermsAggregator.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.bucket.terms;
+
+import com.carrotsearch.hppc.LongLongHashMap;
+import com.carrotsearch.hppc.cursors.LongLongCursor;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.util.CollectionUtil;
+import org.elasticsearch.common.lease.Releasables;
+import org.elasticsearch.common.util.LongHash;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.BucketOrder;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.emptyList;
+
+/**
+ * An aggregator that finds "rare" string values (e.g. terms agg that orders ascending)
+ */
+public class LongRareTermsAggregator extends AbstractRareTermsAggregator<ValuesSource.Numeric, IncludeExclude.LongFilter> {
+
+    static final BucketOrder ORDER = BucketOrder.compound(BucketOrder.count(true), BucketOrder.key(true)); // sort by count ascending
+
+    //TODO review question: is LongLong map ok?
+    protected LongLongHashMap map;
```

Comment:
Is this needed? Since we are calling `collectBucket` and `collectExistingBucket` as we go through collection, can you not ask the bucket for its doc count rather than duplicating it in a separate data structure?

Reply:
I'm still not sure about this because as far as I can see there are the following scenarios when collecting a document:
1. The value in the document is new - we are good because we just add the document to the LongHash and the bucket count is stored as 1.
2. The value in the document has been seen before but is still considered rare (has not hit the threshold doc count) - In this case we are good because we can ask for the doc count using ` bucketDocCount(long bucketOrd)` to determine whether we hit the threshold since the bucketOrd get be found from `bucketOrds`
3. The value in the document has hit the threshold already but the we haven't yet run the merge to delete it from bucketOrds etc. - In this case we will check the filter, determine that the value is not rare and not need to consult the buckets at all
4. The value in the document has hit the threshold already and we have since merged it out - In this case we check the filter, determine that the value is not rare and not need to consult the buckets at all

When we do the merge we recreate all the data structures for the buckets anyway (bucketOrds, doc counts and docDeltas, etc.) so I don't think we need to be able to delete single entries from the bucketOrds? I'm probably missing something here but I'm not sure what?